drivers/sensor/adi/ade9153a/ade9153a.c

```diff
@@ -0,0 +1,556 @@
+/*
+ * Copyright (c) 2024 Plentify (Pty) Ltd.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT adi_ade9153a
+
+#include <zephyr/drivers/sensor.h>
+#include <zephyr/device.h>
+#include <zephyr/init.h>
+#include <zephyr/sys/printk.h>
+#include <zephyr/sys/__assert.h>
+
+#include "ade9153a.h"
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(ade9153a, CONFIG_SENSOR_LOG_LEVEL);
+
+/* Register operations.
+ * The bit[3] = 0 indicates a write action and[3] = 1 indicates a read
+ */
+#define ADE9153A_READ_REG  BIT(3)
+#define ADE9153A_WRITE_REG (0)
+
+static int ade9153a_reg_access(const struct device *dev, uint8_t cmd, uint16_t reg_addr, void *data,
+			       size_t length)
+{
+	const struct ade9153a_config *cfg = dev->config;
+
+	/* The bit [3] = 0 indicates a write action and [3] = 1 indicates a read.
+	 * When reading the cmd value is 8 == BIT(3), when writing cmd is 0
+	 */
+	uint16_t access = (((uint16_t)(reg_addr << 4U) & 0xFFF0U) + cmd);
+
+	access = (access << 8) | ((access >> 8) & 0x00FF);
+
+	const struct spi_buf buf[2] = {{.buf = &access, .len = sizeof(access)},
+				       {.buf = data, .len = length}};
+	struct spi_buf_set tx = {
+		.buffers = buf,
+	};
+
+	if (cmd == ADE9153A_READ_REG) {
+		const struct spi_buf_set rx = {.buffers = buf, .count = 2};
+
+		tx.count = 1;
+
+		return spi_transceive_dt(&cfg->spi_dt_spec, &tx, &rx);
+	}
+
+	tx.count = 2;
+
+	return spi_write_dt(&cfg->spi_dt_spec, &tx);
+}
+
+static inline int ade9153a_get_reg16(const struct device *dev, uint16_t reg_addr, uint16_t *data)
+{
+	uint8_t data_raw[2] = {0};
+
+	int err = ade9153a_reg_access(dev, ADE9153A_READ_REG, reg_addr, data_raw, 2U);
+
+	if (err) {
+		return err;
+	}
+
+	uint16_t data_msb = (((uint16_t)data_raw[0]) << 8U) + (uint16_t)data_raw[1];
+
+	*data = data_msb;
+
+	return 0;
+}
+
+static inline int ade9153a_get_reg32(const struct device *dev, uint16_t reg_addr, uint32_t *data)
+{
+	uint8_t data_raw[4] = {0};
+
+	int err = ade9153a_reg_access(dev, ADE9153A_READ_REG, reg_addr, data_raw, 4U);
+
+	if (err) {
+		return err;
+	}
+
+	uint32_t data_msb = (((uint32_t)data_raw[0]) << 24U) + (((uint32_t)data_raw[1]) << 16U) +
+			    (((uint32_t)data_raw[2]) << 8U) + (uint32_t)(data_raw[3]);
+
+	*data = data_msb;
+
+	return 0;
+}
+static inline int ade9153a_set_reg16(const struct device *dev, uint16_t reg_addr, uint16_t data)
+{
+	uint8_t data_raw[2] = {(uint8_t)data >> 8U, (uint8_t)data};
+
+	return ade9153a_reg_access(dev, ADE9153A_WRITE_REG, reg_addr, data_raw, 2U);
+}
+
+static inline int ade9153a_set_reg32(const struct device *dev, uint16_t reg_addr, uint32_t data)
+{
+	uint8_t data_raw[4] = {(uint8_t)data >> 24U, (uint8_t)data >> 16U, (uint8_t)data >> 8U,
+			       (uint8_t)data};
+
+	return ade9153a_reg_access(dev, ADE9153A_WRITE_REG, reg_addr, data_raw, 4U);
+}
+
+int start_acal(const struct device *dev, int32_t reg_ms_acal_cfg)
```

Comment:
@rodrigopex The argument `reg_ms_acal_cfg` is set as a signed integer but represents an unsigned value being written to a register to configure the type of Msure Acal to run - is this intentional?

Reply:
Changed to `uint32_t`.